Use newer ISO from ipxe for pxeboot

yuvipanda opened the pull request:
Werewulf seems to require bzImage support for compute nodes
pxebooting. However, the default 'network boot' support in
VirtualBox does not seem to support that, nor does it intend
to (https://www.virtualbox.org/ticket/15159). This causes
compute nodes to not come up.

![image](https://user-images.githubusercontent.com/30430/80212313-e1175900-8654-11ea-90ef-3dd2db09618f.png)

Based on [this
comment](https://www.virtualbox.org/ticket/15159#comment:4)
I was able to get compute nodes to come up by using the upstream
IPXE pxeboot image. This is the 'full' version of what VirtualBox
uses, and supports bzImage.

This PR requires now that the user download and keep an ISO
handy. However, we could also do the download ourselves, and spare
the user this extra step.

yuvipanda committed:
Use newer ISO from ipxe for pxeboot

Werewulf seems to require bzImage support for compute nodes
pxebooting. However, the default 'network boot' support in
VirtualBox does not seem to support that, nor does it intend
to (https://www.virtualbox.org/ticket/15159). This causes
compute nodes to not come up.

Based on [this
comment](https://www.virtualbox.org/ticket/15159#comment:4)
I was able to get compute nodes to come up by using the upstream
IPXE pxeboot image. This is the 'full' version of what VirtualBox
uses, and supports bzImage.

This PR requires now that the user download and keep an ISO
handy. However, we could also do the download ourselves, and spare
the user this extra step.

phpHavok commented on Makefile:
```diff
@@ -1,9 +1,10 @@
 NCOMPUTES ?= 2
+PXEBOOT_ISO=
 
 cluster:
```
Could we have this depend on a new target?
```suggestion
$(PXEBOOT_ISO):
    wget -O "$@" http://boot.ipxe.org/ipxe.iso
    
cluster: $(PXEBOOT_ISO)
```

phpHavok commented on Makefile:
```diff
@@ -1,9 +1,10 @@
 NCOMPUTES ?= 2
+PXEBOOT_ISO=
```
Let's set a default of `ipxe.iso`, and we'll use it to create a new target.
```suggestion
PXEBOOT_ISO := ipxe.iso
```

yuvipanda replied:
Set default name of ipxe.iso

phpHavok commented on build-cluster.sh:
```diff
@@ -6,6 +6,12 @@ if [ -z "$NCOMPUTES" ] || [ "$NCOMPUTES" -lt 1 ] || [ "$NCOMPUTES" -gt 10 ]; the
     exit 1
 fi
 
+PXEBOOT_ISO="$2"
+if [ ! -f "$PXEBOOT_ISO" ]; then
+    echo "You need to point to a PXEBOOT ISO from https://ipxe.org/download"
+    exit 1
+fi
+
```
We can remove these checks, as having the target dependency will take care of it.
```suggestion
```

phpHavok reviewed (changes requested):
Hi @yuvipanda! Thanks very much for your pull request. This was working for me out of the box a few months ago, so I'm not sure what changed. Regardless, I appreciate you creating a fix.

I've included some suggested changes below. I'd prefer to download the ipxe.iso file automatically for the user by default, but still allow them to use their own if wanted. If we make the cluster target depend on a new ipxe.iso target, then we can accomplish this. The new target will automatically download the file as part of its recipe, but a user can just use their own file and name it "ipxe.iso" which will satisfy the target and it will skip the download in that case.

Could you make these suggested changes and test that it works?

Thanks so much!

yuvipanda committed:
Make 'cluster' depend on new PXEBOOT_ISO target

Co-Authored-By: Jacob Chappell <chappellind@gmail.com>

yuvipanda committed:
Remove bash code to check for PXEBOOT_ISO

Co-Authored-By: Jacob Chappell <chappellind@gmail.com>

yuvipanda committed:
Set default name of ipxe.iso

Co-Authored-By: Jacob Chappell <chappellind@gmail.com>

yuvipanda committed:
Use wget -c to suppress re-downloading file

This lets users supply their own ipxe.iso file,
and doesn't download them if they match what we
have.

yuvipanda commented:
Thank you  for your suggestions and quick review, @phpHavok! I really appreciate it.

With your changes, I can't get the makefile to work. If I run `make NCOMPUTES=1`, I get:

```
make: Nothing to be done for `ipxe.iso'.
```

If I run `make cluster NCOMPUTES=1`, it makes the cluster dir but doesn't download the iso file.

I've used `wget -c` instead to download file only if needed. This is unideal, but I tried a bunch of things to get Make to work and I couldn't :( If you could help with that, I'm happy to remove the `wget -c`.

Thanks again!

phpHavok commented:
Hi @yuvipanda! Thanks for making the changes. I think you just need to move the new pxeboot target **below** the cluster target. By default, make will use the first target if you don't specify one. So `make NCOMPUTES=1` was actually running on the pxeboot target. (Sorry, my suggestion indicated to place it above, but I wasn't thinking through all the details at the time.) Also, you have to use tabs in Makefiles or it will cause problems (you used spaces to indent).

In short:
1. Revert back to your commit 97154ba1cac114ad8ac23313a013f80896a4b873
2. Move the $(PXEBOOT_ISO) target below the cluster target.
3. Make sure the indentation on the wget line is with a single tab and not with spaces.

Let me know if you need any clarifications. Thanks!

yuvipanda committed:
Revert "Use wget -c to suppress re-downloading file"

Try to get it working with just make again

This reverts commit 3d81eedcc156286f9fc8c405a1edaf6aaecde319.

yuvipanda committed:
Move PXEBOOT_ISO target below cluster

This helps make discover it and operate properly.
Use tabs instead of spaces explicitly

yuvipanda commented:
@phpHavok w00t, that worked! Thank you for the clear guidance! <3

phpHavok reviewed (changes requested):
@yuvipanda I'm glad it worked! I just tested from scratch on my laptop, and found a couple other things that need to be added before merging as it doesn't work out of the box yet. The `build-cluster.sh` uses `$PXEBOOT_ISO` without defining it. Furthermore, we need to copy the ISO file into the `cluster` directory. Please make the following two changes.

First, at the top of `build-cluster.sh` after the `NCOMPUTES=$1` line and its respective if statements, add the following:

```
PXEBOOT_ISO="$2"
if [ ! -f "$PXEBOOT_ISO" ]; then
    echo "You must supply a PXEBOOT ISO that exists."
    exit 1
fi
```

Second, at the very end of `build-cluster.sh`, add the following:

```
cp "$PXEBOOT_ISO" "cluster/$PXEBOOT_ISO"
```

With those two changes, we should be ready to merge!

yuvipanda committed:
Define PXEBOOT_ISO before using it

yuvipanda committed:
Copy PXEBOOT_ISO inside cluster/

yuvipanda commented:
Thanks for the detailed review, @phpHavok!

I've made the changes you suggested...

phpHavok commented:
Sorry for the long delay on this, @yuvipanda . I just tested from scratch, and everything works like a charm now. Commit merged; kudos and many thanks for the contribution!

yuvipanda commented:
@phpHavok woop! Thank you so much for this wonderful project :)
